lib/src/read-contents/read-buffer.js

```diff
@@ -1,23 +1,41 @@
 'use strict';
 
 var fs = require('graceful-fs');
+var iconv = require('iconv-lite');
 var removeBomBuffer = require('remove-bom-buffer');
 
+var getCodec = require('../../codecs');
+
 function bufferFile(file, optResolver, onRead) {
+  var encoding = optResolver.resolve('encoding', file);
+  var codec = getCodec(encoding);
+  if (encoding && !codec) {
+    return onRead(new Error('Unsupported encoding: ' + encoding));
+  }
+
   fs.readFile(file.path, onReadFile);
 
-  function onReadFile(readErr, data) {
+  function onReadFile(readErr, contents) {
     if (readErr) {
       return onRead(readErr);
     }
 
-    var removeBOM = optResolver.resolve('removeBOM', file);
-    if (removeBOM) {
-      file.contents = removeBomBuffer(data);
-    } else {
-      file.contents = data;
+    if (encoding) {
+      var removeBOM = codec.bomAware && optResolver.resolve('removeBOM', file);
+
+      if (codec.enc !== 'utf8') {
+        contents = iconv.decode(contents, encoding, { stripBOM: false });
+        removeBOM = removeBOM && contents[0] === '\ufeff';
+        contents = iconv.encode(contents, 'utf8');
+      }
+
+      if (removeBOM) {
+        contents = removeBomBuffer(contents);
```

Comment:
Yeah, I was doing that in my first stab at this. But then I realized that we'd want to strip BOMs off of UTF16 encoded streams as well, and possibly other multi-byte encodings that support serialization with variable endianness.

So the best way to do that, I thought, was to transcode to UTF8 whilst explicitly _not_ stripping BOMs. At that point, all the various BOM sequences will be mapped to the familiar UTF8 BOM, which we were already equipped to deal with via `remove-bom-buffer` and `-stream`.

But we do take advantage of the decoding/encoding logic understanding about BOMs here: by peeking at the first character of the string in between decode/encode, I can "hint" whether there is a BOM to strip.

Does that make sense?